cli/node/mod.rs

```diff
@@ -0,0 +1,697 @@
+  // for remote modules because they are mainly used to polyfill `node` built
+  // in modules. Another option would be to leave the resolved URLs
+  // as `node:<module_name>` and do the actual remapping to std's polyfill
+  // in module loader. I'm not sure which approach is better.
+  if resolved.scheme().starts_with("http") {
+    return Ok(resolved);
+  }
+
+  // todo(dsherret): cache
+  let encoded_sep_re = Regex::new(r"%2F|%2C").unwrap();
+
+  if encoded_sep_re.is_match(resolved.path()) {
+    return Err(compat::errors::err_invalid_module_specifier(
+      resolved.path(),
+      "must not include encoded \"/\" or \"\\\\\" characters",
+      Some(to_file_path_string(base)),
+    ));
+  }
+
+  let path = to_file_path(&resolved);
+
+  // TODO(bartlomieju): currently not supported
+  // if (getOptionValue('--experimental-specifier-resolution') === 'node') {
+  //   ...
+  // }
+
+  let p_str = path.to_str().unwrap();
+  let p = if p_str.ends_with('/') {
+    p_str[p_str.len() - 1..].to_string()
+  } else {
+    p_str.to_string()
+  };
+
+  let (is_dir, is_file) = if let Ok(stats) = std::fs::metadata(&p) {
+    (stats.is_dir(), stats.is_file())
+  } else {
+    (false, false)
+  };
+  if is_dir {
+    return Err(compat::errors::err_unsupported_dir_import(
+      resolved.as_str(),
+      base.as_str(),
+    ));
+  } else if !is_file {
+    return Err(compat::errors::err_module_not_found(
+      resolved.as_str(),
+      base.as_str(),
+      "module",
+    ));
+  }
+
+  Ok(resolved)
+}
+
+fn module_resolve(
+  specifier: &str,
+  referrer: &ModuleSpecifier,
+  conditions: &[&str],
+  npm_resolver: &dyn DenoDirNpmResolver,
+) -> Result<Option<ModuleSpecifier>, AnyError> {
+  let url = if should_be_treated_as_relative_or_absolute_path(specifier) {
+    let resolved_specifier = referrer.join(specifier)?;
+    Some(resolved_specifier)
+  } else if specifier.starts_with('#') {
+    Some(package_imports_resolve(
+      specifier,
+      referrer,
+      conditions,
+      npm_resolver,
+    )?)
+  } else if let Ok(resolved) = Url::parse(specifier) {
+    Some(resolved)
+  } else {
+    Some(package_resolve(
+      specifier,
+      referrer,
+      conditions,
+      npm_resolver,
+    )?)
+  };
+  Ok(match url {
+    Some(url) => Some(finalize_resolution(url, referrer)?),
+    None => None,
+  })
+}
+
+/// Translates given CJS module into ESM. This function will perform static
+/// analysis on the file to find defined exports and reexports.
+///
+/// For all discovered reexports the analysis will be performed recursively.
+///
+/// If successful a source code for equivalent ES module is returned.
+pub fn translate_cjs_to_esm(
+  file_fetcher: &FileFetcher,
+  specifier: &ModuleSpecifier,
+  code: String,
+  media_type: MediaType,
+  npm_resolver: &GlobalNpmPackageResolver,
+) -> Result<String, AnyError> {
+  fn add_export(source: &mut Vec<String>, name: &str, initializer: &str) {
```

Comment:
I don't see why this should be nested inside this other function. It wouldn't be crazy to have a little unit test for this functionality too.